src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.ForeachToFor;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.ForeachToFor
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpForeachToForCodeRefactoringProvider)), Shared]
+    internal sealed class CSharpForeachToForCodeRefactoringProvider : AbstractForeachToForCodeRefactoringProvider
+    {
+        public CSharpForeachToForCodeRefactoringProvider()
+        {
+        }
+
+        protected override SyntaxNode GetForeachStatement(SyntaxToken token)
+        {
+            var foreachStatement = token.Parent.FirstAncestorOrSelf<ForEachStatementSyntax>();
+            if (foreachStatement == null)
+            {
+                return null;
+            }
+
+            // support refactoring only if caret is in between "foreach" and ")"
+            var scope = TextSpan.FromBounds(foreachStatement.ForEachKeyword.Span.Start, foreachStatement.CloseParenToken.Span.End);
```

Comment:
i would only allow it if it intersects the foreach span.

Reply:
why? I am not sure which other refactoring work that way?

Reply:
For example:  convert if-to-switch:

```c#
                // To prevent noisiness, only show this feature on the 'if' keyword of the if-statement.
                var token = ifStatement.GetFirstToken();
                if (!token.Span.Contains(context.Span))
                {
                    return;
                }
```

And teh reason is as i specified above.  because otherwise it is too noisy.

Reply:
tracking issue - https://github.com/dotnet/roslyn/issues/25667

Reply:
Note: if you do this, can you please show what the experience is if someone goes to the "var" in "foreach (var x in y)" and brings up the lightbulb.  We want to make sure that the "convert to named type" refactoring is above the "convert foreach to for" refactoring.  Note: this is about the refactorings.  not anyhting to do with analyzers.

Reply:
added one more design question - https://github.com/dotnet/roslyn/issues/25667#issuecomment-375503139

now we offering refactoring LB same way we offering code fix, it might be good to add PreferableSpan to Refactoring api so that we can order refactoring as we do for fixers.